iib/workers/tasks/build.py

```
@@ -337,6 +342,152 @@ def _get_resolved_image(pull_spec):
+
+            stdout = get_running_subprocess_output(rpc_proc.stdout)
+            if 'serving registry' in stdout:
+                log.debug('Started the command "%s"', ' '.join(cmd))
+                log.info('Index registry service has been initialized.')
+                return rpc_proc
+
+        rpc_proc.kill()
+
+    raise IIBError(f'Index registry has not been initialized after {max_tries} tries')
+
+
+@retry(attempts=2, wait_on=IIBError, logger=log)
+def _get_present_bundles(from_index, base_dir):
+    """
+    Get a list of bundles already present in the index image.
+
+    :param str from_index: index image to inspect.
+    :param str base_dir: base directory to create temporary files in.
+    :return: list of present bundles as provided by the grpc query.
+    :rtype: list
+    :raises IIBError: if any of the commands fail.
+    """
+    db_path = _get_index_database(from_index, base_dir)
+    port, rpc_proc = _serve_index_registry(db_path)
+
+    bundles = run_cmd(
+        ['grpcurl', '-plaintext', f'localhost:{port}', 'api.Registry/ListBundles'],
+        exc_msg='Failed to get bundle data from index image',
+    )
+    rpc_proc.kill()
+
+    # Transform returned data to parsable json
+    present_bundles = [json.loads(bundle) for bundle in re.split(r'(?<=})\n(?={)', bundles)]
+    return present_bundles
+
+
+@retry(attempts=2, wait_on=IIBError, logger=log)
+def _get_missing_bundles(present_bundles, bundles):
+    """
+    Filter out bundles to only those not present in the index image.
+
+    :param list present_bundles: list of bundles present in the index image, as provided by opm.
+    :param list bundles: resolved bundles requested to be added to the index image.
+    :return: list of bundles not present in the index image.
+    :rtype: list
+    """
+    present_bundle_hashes = []
+    filtered_bundles = []
+    for bundle in present_bundles:
+        if '@sha256:' not in bundle['bundlePath']:
+            raise IIBError(
```

I don't think we should raise an error here if the index image contains a bundle image specified by tag. I would recommend removing this check and simply skipping that bundle

---

The reason I say this is if the bundle with tag clashes with a bundle image that we have a sha for, OPM will anyways throw an error. So let OPM handle that.

---

Removed.